backstack/src/commonMain/kotlin/com/slack/circuit/backstack/SaveableBackStack.kt

```diff
@@ -46,25 +50,71 @@ public class SaveableBackStack : BackStack<SaveableBackStack.Record> {
   public override val topRecord: Record?
     get() = entryList.firstOrNull()
 
-  public override fun push(screen: Screen) {
-    push(screen, emptyMap())
+  public override fun push(screen: Screen, resultKey: String?) {
+    return push(screen, emptyMap(), resultKey)
   }
 
-  public fun push(screen: Screen, args: Map<String, Any?>) {
-    push(Record(screen, args))
+  public fun push(screen: Screen, args: Map<String, Any?>, resultKey: String?) {
+    push(Record(screen, args), resultKey)
   }
 
-  public override fun push(record: Record) {
+  public override fun push(record: Record, resultKey: String?) {
     entryList.add(0, record)
+    // Clear the cached pending result from the previous top record
+    // TODO this may not be enough. See TODO in prepareForResult()
+    entryList.getOrNull(1)?.apply { resultKey?.let(::prepareForResult) }
   }
 
-  override fun pop(): Record? = entryList.removeFirstOrNull()
+  override fun pop(result: PopResult?): Record? {
+    val popped = entryList.removeFirstOrNull()
+    result?.let {
+      // Send the pending result to our new top record
+      topRecord?.sendResult(it)
+    }
+    return popped
+  }
 
   public data class Record(
     override val screen: Screen,
     val args: Map<String, Any?> = emptyMap(),
     override val key: String = uuid4().toString(),
   ) : BackStack.Record {
+    /**
+     * A [Channel] of pending results. Note we use this instead of a [CompletableDeferred] because
+     * we may push and pop back to a given record multiple times, and thus need to be able to push
+     * and receive multiple results.
+     *
+     * TODO what's the right behavior here?
+     * - Capacity 1 + overflow drop oldest: we only care about the most recent result
```

Comment:
Not initially sure we want the most recent result, might be some issue with dropping results 🤔  

In that case I wonder if Capacity 1 + suspend works?

Reply:
I don't think so, for a couple reasons.

1. The technical reason is we never send, only trySend. As such, it would never be able to suspend.
2. The practical reason is I don't think this situation could ever happen given that we clear the result on a new launch if it's unconsumed, so there shouldn't ever be a case where we could have a second result pushed back here. The channel at this point really just becomes more of a reusable CompletableDeferred that allows easy try* access.